projects/dendrites/permutedMNIST/run_dendritic_network.py

```diff
@@ -0,0 +1,231 @@
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+"""
+This module trains & evaluates a dendritic network in a continual learning setting on
+permutedMNIST for a specified number of tasks/permutations. A context vector is
+provided to the dendritic network, so task information need not be inferred.
+
+This setup is very similar to that of context-dependent gating model from the paper
+'Alleviating catastrophic forgetting using contextdependent gating and synaptic
+stabilization' (Masse et al., 2018).
+"""
+
+import math
+from collections import defaultdict
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+
+from nupic.research.frameworks.dendrites import DendriticMLP
+from nupic.research.frameworks.pytorch.datasets import ContextDependentPermutedMNIST
+from nupic.research.frameworks.vernon import ContinualLearningExperiment, mixins
+
+
+# ------ Experiment class
+class PermutedMNISTExperiment(mixins.RezeroWeights,
+                              ContinualLearningExperiment):
+
+    def setup_experiment(self, config):
+        # Store optimizer args, since optimizer is reset at the beginning of each task
+        self.optimizer_class = config.get("optimizer_class", torch.optim.SGD)
+        self.optimizer_args = config.get("optimizer_args", {})
+
+        super().setup_experiment(config)
+
+    @classmethod
+    def compute_task_indices(cls, config, dataset):
+        # Assume dataloaders are already created
+        class_indices = defaultdict(list)
+        for idx, (_, _, target) in enumerate(dataset):
+            class_indices[target].append(idx)
+
+        # Defines how many classes should exist per task
+        num_tasks = config.get("num_tasks", 1)
+        num_classes = config.get("num_classes", None)
+        assert num_classes is not None, "num_classes should be defined"
+        num_classes_per_task = math.floor(num_classes / num_tasks)
+
+        task_indices = defaultdict(list)
+        for i in range(num_tasks):
+            for j in range(num_classes_per_task):
+                task_indices[i].extend(class_indices[j + (i * num_classes_per_task)])
+        return task_indices
+
+    def run_iteration(self):
+        # Reset optimizer before training on a new task, just as context-dependent
```

Comment:
Is this ever called in the script? I couldn't see where.

Reply:
I removed it, and placed the optimizer reset within the continual learning loop.